search/typesense.py

```diff
@@ -0,0 +1,538 @@
+- The Courseware content search in the Learning MFE
+  (gated by the `courseware.mfe_courseware_search` waffle flag).
+
+Other use cases are not supported.
+
+To switch to this engine, you should run a Typesense instance and define the
+following setting:
+
+    SEARCH_ENGINE = "search.typesense.TypesenseEngine"
+
+You will then need to create the new indices by running:
+
+    ./manage.py lms shell -c "import search.typesense; search.typesense.create_indexes()"
+
+^ This manual index creation is a temporary approach, while we're evaluating
+Typesense. A proper long-term approach is needed for all our search engines, as
+discussed at https://github.com/openedx/edx-platform/issues/36868 .
+
+For more information about the Typesense API in Python, check
+https://github.com/typesense/typesense-python
+"""
+import logging
+import typing as t
+
+from django.conf import settings
+from django.utils import timezone
+import typesense
+from typesense.collection import Collection
+from typesense.exceptions import ObjectNotFound, RequestMalformed, ServerError
+from typesense.types.collection import RegularCollectionFieldSchema
+from typesense.types.document import Hit, SearchResponse
+
+from search.search_engine_base import SearchEngine
+from search.utils import (
+    convert_doc_datatypes,
+    IS_NULL_SUFFIX,
+    restore_doc_datatypes,
+    UTC_OFFSET_SUFFIX,
+    ValueRange,
+)
+
+# ====== Indices ======
+# Due to the messy architecture of edx-search, we have to encode information about specific indexes in this engine.
+# So some information about the indexes is here (in ES, Meilisearch, and Typesense engines separately) and other info
+# is in edx-platform, e.g. edx-platform/cms/djangoapps/contentstore/courseware_index.py
+#
+# The two indexes we're most concerned about are:
+#  - Course Info index: stores a list of courses that users can browse and then enroll into
+#  - Course Content index: stores the course content (XBlocks) for enrolled users to search within one course
+# We need to know which fields store timestamps in order to force them to be stored as 64 bit integers instead of 32
+COURSE_INFO_INDEX = getattr(settings, "COURSEWARE_INFO_INDEX_NAME", "course_info")
+COURSE_CONTENT_INDEX = getattr(settings, "COURSEWARE_CONTENT_INDEX_NAME", "courseware_content")
+INDEX_CONFIGURATION = {
+    COURSE_INFO_INDEX: {
+        # Most fields will be auto-created but some (datetimes, facet fields) need to be specified up front:
+        "field_overrides": [
+            {"name": "start", "type": "datetime", "optional": True},
+            {"name": "end", "type": "datetime", "optional": True},
+            {"name": "enrollment_start", "type": "datetime", "optional": True},
+            {"name": "enrollment_end", "type": "datetime", "optional": True},
+            {"name": "org", "type": "string", "facet": True},
+            {"name": "modes", "type": "string[]", "facet": True},
+            {"name": "language", "type": "string", "facet": True},
```

Comment:
I discovered courses in the wild that did not have a language field and failed to index:

```
2025-10-20 04:16:43,829 ERROR 414 [search.typesense] [user None] [ip None] typesense.py:142 - Failed to index document REDACTED: Field `language` has been declared in the schema, but is not found in the document.
```

I guess this should be an optional field:

```suggestion
            {"name": "language", "type": "string", "facet": True, "optional": True},
```